### ISSUE TITLE
[Request] Indexed color support for CGFX

### ISSUE THREAD

**Issue by frozenLake** [shonumi/gbe-plus#71]
This probably can be saved for after PNG support is implemented, but there is also indexed color for BMP, so I'll just continue.

Anyway, by having indexed color support, we could have something like

MegamanSheet.bmp
00s = 000000, 000000, 0070E8, 00E8D8, F8E0A0, F8F8F8, 00FF00
00o = A80010, 000000, 0070E8, 00E8D8, F8E0A0, F8F8F8, 00FF00
00k = E00058, 000000, 0070E8, 00E8D8, F8E0A0, F8F8F8, 00FF00
00g = F870B0, 000000, 0070E8, 00E8D8, F8E0A0, F8F8F8, 00FF00
01w = 5890F8, 000000, 0070E8, 00E8D8, F8E0A0, F8F8F8, 00FF00
02Q = 00E8D8, 000000, 000000, 0070E8, F8E0A0, F8F8F8, 00FF00
038 = 0070E8, 000000, 00E8D8, 000000, F8E0A0, F8F8F8, 00FF00

instead of having seven different files for each palette swap while charging.

**Comment by shonumi** [comment]
I do not think this would be a good approach due to the complexity it adds to parsing BMPs (essentially having 2 code paths, 1 for normal 24-bit color images, and another for indexed color ones).

Instead, what you want is something I've experimented with called the **EXT_AUTO_PAL** option, which would be another option users could enable like the **EXT_AUTO_VRAM_ADDR** and **EXT_AUTO_BRIGHT**. It uses an algorithm to automatically swap colors when the same tile uses different palettes. That feature is still WIP though.

At any rate, this is the preferred solution since it's relatively easy on the codebase and still requires users to dump only 1 version of their graphics (as long as they enable the option). It would require a new manifest file format going forward, but I've working on making a new format for a while now, so that change is inevitable.

The issue of DMG palette changes has been a known thorn in my side for quite some time though. It greatly affects other games (e.g. games that change the palette when paused, and dozens of Pokemon attacks like Night Shade), so it's very much something I'm trying to fix.

**Comment by frozenLake** [comment]
Actually, couldn't we encode what colors are the "indexes" in a line? For instance, have the default colors be for 00o, and then remap color matches from that, so that we have...

0=A80010
1=0070E8
2=00E8D8

Note that I am only picking specific colors to be indexed, as the others don't get remapped when charging. However, it might come up if we are dealing with colors when fading, so, one might need to index those anyway.

but yeah, changing specific hex values to other hex values, maybe?

**Comment by frozenLake** [comment]
So with the way EXT_AUTO_PAL is described, it seems that I won't be able to duplicate the NES Megaman charging for the gameboy games.

It also seems that using this with having more colors than the original gameboy color palette will be... weird. Either way, both are cases that would take duplicating the image to solve.

Its a shame that the end user can't define custom palettes for palettes, but completely understandable from a coding perspective.

EXT_AUTO_PAL does seem like something that will be fun to play around with, though!
